rfcs/OptInFeatures.md

```diff
@@ -0,0 +1,152 @@
+
+### Non-goals
+This proposal is not about security and/or hiding parts of a schema. Its goal is to make it easier to communicate opt-in features to client developer and therefore needs to expose that information.
+
+## 👀 Prior work
+
+* GitHub uses [schema previews](https://docs.github.com/en/graphql/overview/schema-previews) to opt-in new features.
+* Kotlin has [OptIn requirements](https://kotlinlang.org/docs/opt-in-requirements.html) that started out as `@Experimental` before [being changed to `@OptIn`](https://youtrack.jetbrains.com/issue/KT-26216/Generalize-Experimental-API)
+
+## 🧑‍💻 Proposed solution
+
+### The `@optIn` directive
+
+It is proposed to add an `@optIn` directive to the specification:
+
+```graphql
+"""
+Indicates that the given field, argument, input field or enum value requires
+giving explicit consent before being used.
+"""
+directive @optIn(feature: String!) repeatable
+    on FIELD_DEFINITION 
+    | ARGUMENT_DEFINITION 
+    | INPUT_FIELD_DEFINITION 
+    | ENUM_VALUE
+```
+
+The `optIn` directive can then be used in the schema. For an example, to signal an experimental field:
+
+```graphql
+type Session {
+  id: ID!
+  title: String!
+  # [...]
+  startInstant: Instant @optIn(feature: "experimentalInstantApi")
+  endInstant: Instant @optIn(feature: "experimentalInstantApi")
+}
+```
+
+### Introspection
+
+> This section is a proposal based on the current introspection mechanism. A more global mechanism (see [#300](https://github.com/graphql/graphql-spec/issues/300)) would make it obsolete
+
+`@optIn` features should be hidden from introspection by default and include if `includeOptIn` contains the given feature:
+
+```graphql
+type __Type {
+  kind: __TypeKind!
+  name: String
+
+  # [...] other fields omitted for clarity
+
+  # includeOptIn is a list of features to include
+  fields(includeDeprecated: Boolean = false, includeOptIn: [String!]): [__Field!]  
+}
+```
+
+Tools can get a list of `@optIn` features required to use a field (or input field, argument, enum value) using `requiresOptIn`:
+```graphql
+type __Field {
+  name: String!
+  isDeprecated: Boolean!
+
+  # [...] other fields omitted for clarity
+
+  # list of @optIn features required to use this field
+  requiresOptIn: [String!]
+  args(includeDeprecated: Boolean = false, includeOptIn: [String!]): [__InputValue!]!
+}
+```
+
+A given field is included in introspection results if all the conditions are satisfied. In pseudo code, if the following condition is true:
+
+```
+includeOptIn.containsAll(field.requiresOptIn) && (includeDeprecated || !field.isDeprecated)
+```
+
+### Validation
+
+Similarly to [deprecation](https://spec.graphql.org/draft/#sel-FAHnBZNCAACCwDqvK), the `@optIn` directive must not appear on required (non-null without a default) arguments or input object field definitions.
+
+In other words, `@optIn`  arguments or input fields, must be either nullable or have a default value.
+
+## 🗳️ Alternate solutions
```

Comment:
### Alternative proposal: @requiresOptIn and @optIn

Same as the _Proposed solution_, except the schema uses `@requiresOptIn` instead
```graphql
directive @requiresOptIn(feature: String!) on FIELD_DEFINITION 
```
```graphql
type Session {
  id: ID!
  title: String!
  startInstant: Instant @requiresOptIn(feature: "experimentalInstantApi")
  endInstant: Instant @requiresOptIn(feature: "experimentalInstantApi")
}
```
Operations use `@optIn`
```graphql
directive @optIn(feature: String!) on FIELD 
```
```graphql
query {
  session {
    id
    title
    startInstant @optIn(feature: "experimentalInstantApi")
    endInstant @optIn(feature: "experimentalInstantApi")
  }
}
```
Adding `@requiresOptIn` to an existing field would need to be detected as a schema breaking change.
Removing `@requiresOptIn` is safe, and an operation using `@optIn` on a field without `@requiresOptIn` would continue to work.
Tools can use introspection values to provide code actions to add `@optIn` where necessary
Matching the `feature` prevents misuse, for example:
```graphql
query {
  session @optIn {
    id @optIn
    title @optIn
    startInstant @optIn
    endInstant @optIn
  }
}
```

For introspection, unlike the _Proposed solution_ the schema is telling you about the opt-in features, rather than you having to provide them
```diff
 type __Type {
   # [...] other fields omitted for clarity
   fields(
     includeDeprecated: Boolean = false,
+    includeRequiresOptIn: Boolean = false
   ): [__Field!]
 }
 
 type __Field {
   name: String!
   # [...] other fields omitted for clarity
   isDeprecated: Boolean!
   deprecationReason: String
+  requiresOptIn: Boolean!
+  optInFeature: String
 }
```

Pros:
  - Self-documenting. There's no set of feature names defined somewhere else
 - Usage is colocated with consumption
 - No grammar changes
 - No need to enumerate _all opt in features_

Cons:
 - Only uses a string, no capability for providing more type information
 - Goes further than just being documentation to somewhat imply behaviour (although that could still be for server to decide)
  - No place to document a feature itself (i.e one feature might apply to many fields, and want a high-level description of the feature itself)

Reply:
> Alternative proposal: `@requiresOptIn`

Woops, sorry I missed this earlier today, I actually just pushed `@requiresOptIn`, I like this a lot, it plays well with `@optIn` and is more explicit

> Operations use `@optIn`

Do we need to put that in the spec? This has execution implication and different teams might want to implement that differently. 

I initially envisioned the proposal as a way to "communicate" opt-in requirements (like `@deprecated` communicates the deprecated status). "Enforcing" it seems another steps with a wide range of implications.

> For introspection, unlike the Proposed solution the schema is telling you about the opt-in features, rather than you having to provide them

Not sure I get this one. The current proposed solution already exposes the opt-in features:

```graphql
type __Field {
  name: String!
  isDeprecated: Boolean!

  # [...] 

  # list of @requiresOptIn features required to use this field
  requiresOptIn: [String!]
```

We can use an additional Boolean but it feels like duplication because we can tell if there's the need for duplication if the list is not empty. 

On a separate (but related) note, I wish `isDeprecated` would be replaced by checking for `deprecationReason != null`.